### PR TITLE
HideProcessAndCopyOnReadAndBlockAccessWMITemplateAndExtendTempTemplateAndUiForLimitAndNewBoxTypeAndBlockLocalConnectTemplate

### PR DESCRIPTION
**Pull request by love-code-yeyixiao**
1. Add feature to hide unsandboxed processes.(Successful)
2. Add feature "CopyOnOpen" in order to migrate files while it only read a file.(Wait fixing,but not affect programs now,so it could merge into master branch.I have no idea fixing it up at this time.)
3. Add template "BlockAccessWMI" to avoid program getting system deatils or bypass hiding processes through WMI.
4. Extend the feature which delete temportry templates automantic when a box closed,make it could delete Local template refernce and itself(The latter didn't work well yet,it delete refernces but not include section in Sandboxie.ini)
5. Add UI for memory and number limits.
6. Add new box type "Maximize isolation sandbox"
7. Add new template "Block Local Connect"
8. Add code for issue #3975

### CODE REVIEW

**Comment by offhub** [comment]
How does the `HideInteractionProcess` setting differ from the `HideHostProcess` setting?

**Comment by love-code-yeyixiao** [comment]
> How does the `HideInteractionProcess` setting differ from the `HideHostProcess` setting?

The latter requires the process name to be specified manually, but this setting is intended to override possible user processes in all environments with only one setting.

File: Sandboxie/core/dll/sysinfo.c [hunk 1]
```diff
@@ -206,6 +206,37 @@ _FX NTSTATUS SysInfo_NtQuerySystemInformation(
 
         SysInfo_DiscardProcesses(Buffer);
     }
+	if (NT_SUCCESS(status) && (SystemInformationClass == SystemFirmwareTableInformation) && SbieApi_QueryConfBool(NULL, "HideFirmWareInfo", FALSE)) {
```
Comment (offhub) [inline Sandboxie/core/dll/sysinfo.c:209]:
```suggestion
	if (NT_SUCCESS(status) && (SystemInformationClass == SystemFirmwareTableInformation) && SbieApi_QueryConfBool(NULL, L"HideFirmwareInfo", FALSE)) {
```

File: Sandboxie/core/dll/sysinfo.c [hunk 2]
```diff
@@ -206,6 +206,37 @@ _FX NTSTATUS SysInfo_NtQuerySystemInformation(
...
+				 LPDWORD lpReserved,
+				 LPDWORD lpType,
+				 LPBYTE lpData,
+				LPDWORD lpcbData
+			);
+		ROK RegOpenKeyExW=Ldr_GetProcAddrOld(L"Advapi32.dll",L"RegOpenKeyExW");
+		RQVEW RegQueryValueExW = Ldr_GetProcAddrOld(L"Advapi32.dll", L"RegQueryValueExW");
+		DWORD type;
+		if (RegOpenKeyExW(HKEY_CURRENT_USER, L"SOFTWARE\\SandboxieHide\\", 0, KEY_READ, &hKey))
+			RegQueryValueExW(hKey, "FalseFirewareValue", 0, &type, lpData, &dwLen);
```
Comment (offhub) [inline Sandboxie/core/dll/sysinfo.c:234]:
```suggestion
			RegQueryValueExW(hKey, "FalseFirmwareValue", 0, &type, lpData, &dwLen);
```

File: Sandboxie/core/dll/sysinfo.c [hunk 3]
```diff
@@ -261,25 +334,33 @@ _FX void SysInfo_DiscardProcesses(SYSTEM_PROCESS_INFORMATION *buf)
         next = (SYSTEM_PROCESS_INFORMATION *) (((UCHAR *)curr) + curr->NextEntryOffset);
         if (next == curr)
             break;
-
-		SbieApi_QueryProcess(next->UniqueProcessId, boxname, NULL, NULL, NULL);
-
+		SbieApi_QueryProcess(next->UniqueProcessId, boxname, NULL, tempSid, &tempSession);
 		BOOL hideProcess = FALSE;
+		if(_wcsnicmp(tempSid, L"S-1-5-18",8) != 0 && _wcsnicmp(tempSid, L"S-1-5-80",8) != 0 /* && _wcsnicmp(tempSid, L"S-1-5-20", 8) != 0 && _wcsnicmp(tempSid, L"S-1-5-6", 7) != 0 */ && SbieApi_QueryConfBool(NULL, L"HideNonSystemProcess", FALSE)) {
```
Comment (offhub) [inline Sandboxie/core/dll/sysinfo.c:339]:
```suggestion
		if(_wcsnicmp(tempSid, L"S-1-5-18",8) != 0 && _wcsnicmp(tempSid, L"S-1-5-80",8) != 0 /* && _wcsnicmp(tempSid, L"S-1-5-20", 8) != 0 && _wcsnicmp(tempSid, L"S-1-5-6", 7) != 0 */ && SbieApi_QueryConfBool(NULL, L"HideNonSystemProcesses", FALSE)) {
```

File: Sandboxie/core/dll/sysinfo.c [hunk 4]
```diff
@@ -261,25 +334,33 @@ _FX void SysInfo_DiscardProcesses(SYSTEM_PROCESS_INFORMATION *buf)
...
 				for (hiddenProcessesPtr = hiddenProcesses; *hiddenProcessesPtr != L'\0'; hiddenProcessesPtr += wcslen(hiddenProcessesPtr) + 1) {
 					if (_wcsicmp(imagename, hiddenProcessesPtr) == 0) {
 						hideProcess = TRUE;
 						break;
 					}
 				}
 			}
+			if (!hideProcess) {
+				if (_wcsnicmp(imagename, L"Sandboxie", 9) == 0 || _wcsnicmp(imagename, L"Sbie", 4) == 0) {
+					if (SbieApi_QueryConfBool(NULL, L"HideSbieProcess", FALSE))
```
Comment (offhub) [inline Sandboxie/core/dll/sysinfo.c:360]:
```suggestion
					if (SbieApi_QueryConfBool(NULL, L"HideSbieProcesses", FALSE))
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - Add UI option to start unsandboxed process but force child processes in SelectBoxWindow.
 - Add option "AlertBeforeStart".When it is set,a prompt pops up before launching a new program into the sandbox using "Start.exe" and checks if the program that started "Start.exe" is a Sandboxie component itself,if it is not, a warning pops up.
 - Add option for EditAdminOnly in SetupWizard.
+- Add option "HideNonSystemProcess" to hide processes other than some system services from processes lists for sandboxed processes.
```
Comment (offhub) [inline CHANGELOG.md:17]:
```suggestion
- Add option "HideNonSystemProcesses" to hide processes other than some system services from processes lists for sandboxed processes.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+- Add option "HideSbieProcess" to hide Sandboxie Work Process(SbieSvc,SandboxieRpcSs,etc.).
```
Comment (offhub) [inline CHANGELOG.md:18]:
```suggestion
- Add option "HideSbieProcesses" to hide Sandboxie Work Process(SbieSvc,SandboxieRpcSs,etc.).
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+- When "HideFirmWareInfo" is set,the programs which try getting fireware info will get false data from HKEY_CURRENT_USER\\SOFTWARE\\SandboxieHide\\FalseFirewareValue
```
Comment (offhub) [inline CHANGELOG.md:19]:
```suggestion
- When "HideFirmwareInfo" is set,the programs which try getting fireware info will get false data from HKEY_CURRENT_USER\\SOFTWARE\\SandboxieHide\\FalseFirmwareValue
```

File: Sandboxie/install/Templates.ini [hunk 8]
```diff
@@ -3771,6 +3771,12 @@ Tmpl.Title=Block DNS
 Tmpl.Class=Misc
 NetworkAccess=*,Block;Port=53;Protocol=UDP
 
+[Template_BlockAccessWMI]
+Tmpl.Title=Block Accessing WMI
+Tmpl.Class=Misc
+ClosedFilePath=%SystemRoot%\System32\wbem\*
+
+
```
Comment (offhub) [inline Sandboxie/install/Templates.ini:3779]:
```suggestion

```

File: Sandboxie/install/Templates.ini [hunk 9]
```diff
@@ -3771,6 +3771,16 @@ Tmpl.Title=Block DNS
...
 
+[Template_BlockAccessWMI]
+Tmpl.Title=Block Accessing WMI
+Tmpl.Class=Misc
+ClosedFilePath=%SystemRoot%\System32\wbem\*
+
+[Template_BlockLocalConnect]
+Tmpl.Title=Block Local Connect
+Tmpl.Class=Misc
+NetworkAccess=*,Block;Port=*;Address=*;Protocol=Any
```
Comment (offhub) [inline Sandboxie/install/Templates.ini:3782]: This setting will block all connections. If this is the intention, you can remove the part after Block.

```suggestion
NetworkAccess=*,Block
```

File: SandboxiePlus/SandMan/SbiePlusAPI.cpp [hunk 10]
```diff
@@ -813,6 +813,21 @@ CSandBoxPlus::EBoxTypes CSandBoxPlus::GetTypeImpl() const
 	if (m_iUnsecureDebugging != 0)
 		return eInsecure;
 
+	BOOL isMAX = GetBool("HideNonSystemProcess") &&
```
Comment (offhub) [inline SandboxiePlus/SandMan/SbiePlusAPI.cpp:816]:
```suggestion
	BOOL isMAX = GetBool("HideNonSystemProcesses") &&
```

File: SandboxiePlus/SandMan/Windows/OptionsGeneral.cpp [hunk 11]
```diff
@@ -1044,8 +1060,24 @@ void COptionsWindow::UpdateBoxType()
 	bool bPrivacyMode = ui.chkPrivacy->isChecked();
 	bool bSecurityMode = ui.chkSecurityMode->isChecked();
 	bool bAppBox = ui.chkNoSecurityIsolation->isChecked();
+	bool bIsoationMax = m_pBox->GetBool("HideNonSystemProcess")
+		&& m_pBox->GetBool("HideNonSystemProcess")
```
Comment (offhub) [inline SandboxiePlus/SandMan/Windows/OptionsGeneral.cpp:1064]:
```suggestion
		&& m_pBox->GetBool("HideNonSystemProcesses")
```

File: Sandboxie/install/Templates.ini [hunk 12]
```diff
@@ -3779,7 +3779,7 @@ ClosedFilePath=%SystemRoot%\System32\wbem\*
 [Template_BlockLocalConnect]
 Tmpl.Title=Block Local Connect
 Tmpl.Class=Misc
-NetworkAccess=*,Block;Port=*;Address=*;Protocol=Any
+NetworkAccess=*,Block;Port=*;Address=127.*.*.*;Protocol=Any
```
Comment (offhub) [inline Sandboxie/install/Templates.ini:3782]:
```suggestion
NetworkAccess=*,Block;Port=*;Address=127.0. 0.0-127.255. 255.255;Protocol=Any
```

**Comment by offhub** [comment]
1. This setting also hides system processes outside the sandbox. (is it bug?)
5. It's not functional.

**Pull request merged**